### PR TITLE
[pvr] Support PVR Recording ID in JSON Playlist.Add and Playlist.Insert API

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Add support for `recordingid` in the `Playlist.Add` and `Playlist.Insert` JSON API calls.

## Motivation and context

Because playcount cannot work for PVR recordings unless we do so for Remote Apps. Investigation here: https://forum.kodi.tv/showthread.php?tid=364983&pid=3078870#pid3078870

## How has this been tested?

Waiting iOS remote app maintainer to test.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by ksooo — COMMENTED** [review]
Is there no other way than to introduce a hard PVR dependency to the VideoLibrary? Feels ugly if not wrong.

**Comment by ksooo** [comment]
This introduces a circular dependency between PVR recordings and video database. Currently PVR recordings uses Video database, now we're going in circles as video database uses PVR recordings. Nothing I can approve, sorry.

**Comment by phunkyfish** [comment]
If it works I’ll find a better way.

File: xbmc/interfaces/json-rpc/VideoLibrary.cpp [hunk 1]
```diff
@@ -1079,6 +1083,21 @@ bool CVideoLibrary::FillFileItemList(const CVariant &parameterObject, CFileItemL
...
+    if (CServiceBroker::GetPVRManager().IsStarted())
+    {
+      std::shared_ptr<PVR::CPVRRecordings> recordings =
+          CServiceBroker::GetPVRManager().Recordings();
+      if (recordings)
+      {
+        const std::shared_ptr<PVR::CPVRRecording> recording =
+            recordings->GetById(static_cast<int>(parameterObject["recordingid"].asInteger()));
+        if (recording)
+          list.Add(std::make_shared<CFileItem>(recording));
```
[the remark this answers is not in the document]
Comment (phunkyfish) [reply xbmc/interfaces/json-rpc/VideoLibrary.cpp:1098]: Fixed

**Comment by phunkyfish** [comment]
> This introduces a circular dependency between PVR recordings and video database. Currently PVR recordings uses Video database, now we're going in circles as video database uses PVR recordings. Nothing I can approve, sorry.

How about a public getter for the `FileItemPtr` in `PVROperations`?

**Comment by ksooo** [comment]
> This introduces a circular dependency between PVR recordings and video database.

I was wrong with this, mixed up `CVideoDatabase` and `CVideoLibrary`, sorry. I still would like if we could find a solution without introducing compile time dependency to PVR here, but this is not a blocker.

As I cannot currently find a better solution, it would be fine for me if this PR gets merged as it is.

File: xbmc/interfaces/json-rpc/VideoLibrary.cpp [hunk 2]
```diff
@@ -1079,6 +1083,24 @@ bool CVideoLibrary::FillFileItemList(const CVariant &parameterObject, CFileItemL
       success = true;
     }
   }
+  if (recordingID > 0)
+  {
+    if (CServiceBroker::GetPVRManager().IsStarted())
+    {
+      std::shared_ptr<PVR::CPVRRecordings> recordings =
```
Comment (ksooo) [inline xbmc/interfaces/json-rpc/VideoLibrary.cpp:1090]: const std::...

**Review by ksooo — APPROVED** [review]

File: xbmc/interfaces/json-rpc/VideoLibrary.cpp (hunk 2 above)
Comment (phunkyfish) [reply xbmc/interfaces/json-rpc/VideoLibrary.cpp:1090]: Done

**Comment by phunkyfish** [comment]
> As I cannot currently find a better solution, it would be fine for me if this PR gets merged as it is.

I think encapsulating it in PVROperations class is the cleanest way to do this. Modified the PR to do it this way.

**Review by howie-f — APPROVED** [review]
did a successful test https://forum.kodi.tv/showthread.php?tid=364983&pid=3078949#pid3078949

but i'm not sure which commit it was (https://github.com/xbmc/xbmc/commit/67bd6c983e962bdde357bc02c0f20676903b6992 i think)

**Comment by phunkyfish** [comment]
> did a successful test https://forum.kodi.tv/showthread.php?tid=364983&pid=3078949#pid3078949
> 
> but i'm not sure which commit it was ([67bd6c9](https://github.com/xbmc/xbmc/commit/67bd6c983e962bdde357bc02c0f20676903b6992) i think)

Cool, thanks. If we can test the playcount increments correctly after using the new method that would be cool.

**Comment by phunkyfish** [comment]
@DaveTBlake as this touches JSON would you take a look? As this is an JSON API change I assume a backport is not possible.

**Comment by garbear** [comment]
> @DaveTBlake as this touches JSON would you take a look? As this is an JSON API change I assume a backport is not possible.

Patch increments can be backported IMO. Otherwise any problems in the API would persist indefinitely.

**Comment by phunkyfish** [comment]
> > @DaveTBlake as this touches JSON would you take a look? As this is an JSON API change I assume a backport is not possible.
> 
> Patch increments can be backported IMO. Otherwise any problems in the API would persist indefinitely.

Ok, let me see what @DaveTBlake thinks

File: xbmc/interfaces/json-rpc/schema/version.txt [hunk 3]
```diff
@@ -1 +1 @@
-JSONRPC_VERSION 12.5.0
+JSONRPC_VERSION 12.5.1
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/schema/version.txt:1]: As a non-breaking addition to the API this is a minor version bump e.g. 12.7.0

**Review by DaveTBlake — COMMENTED** [review]
> As this is an JSON API change I assume a backport is not possible.

Non-breaking API changes can be backported right up until we make changes to API that we don't backport. That has now happened, so no this can not be backported. I did hold off the API in Matrix and Nexus getting out of sync for along as reasonable, but we are there.

>Patch increments can be backported IMO. Otherwise any problems in the API would persist indefinitely.

@garbear you need to consider how the JSON API versioning is used by clients. The versioning has always been hierarchical, so cleints know that from version x.x.x onwards it has certain properties (features, functions, or fixes), and not a jumble. It is not my design but how API has been implemented. So no, _once API in master diverges_ fixes don't get backported and have to wait for the new release.

File: xbmc/interfaces/json-rpc/schema/version.txt (hunk 3 above)
Comment (DaveTBlake) [reply xbmc/interfaces/json-rpc/schema/version.txt:1]: Ha yes 12.6.0, I obviously can't count this morning!
To be fair I had also been looking at another API change PR so may have bumped it in my head :)

Comment (DaveTBlake) [reply xbmc/interfaces/json-rpc/schema/version.txt:1]: Merged that PR so 12.7.0 is right now after rebase.

**Comment by phunkyfish** [comment]
@wutschel FYI.

**Review by ksooo — APPROVED** [review]
LGTM

**Pull request merged**